src/trufflesom/interpreter/nodes/bc/BytecodeLoopNode.java

```diff
@@ -631,12 +640,14 @@ public Object executeGeneric(final VirtualFrame frame) {
         }
 
         case RETURN_LOCAL: {
-          LoopNode.reportLoopCount(this, backBranchesTaken);
-          return stack[stackPointer];
+          LoopNode.reportLoopCount(this, c.backBranchesTaken);
+          returnValue = stack[stackPointer];
+          nextBytecodeIndex = RETURN_FROM_METHOD;
+          break;
```

Comment:
You're still breaking here, maybe that makes a different. In TruffleSqueak, we `continue` (see https://github.com/hpi-swa/trufflesqueak/blob/23e54fbbbe3738d48eb93086ee4d21643c835caa/src/de.hpi.swa.trufflesqueak/src/de/hpi/swa/trufflesqueak/nodes/ExecuteBytecodeNode.java#L136-L137).

Reply:
The only thing after the `break` is the `bytecodeIndex = nextBytecodeIndex;` at the [bottom of the loop](https://github.com/SOM-st/TruffleSOM/pull/157/files#diff-b8269bcc854c3ebc1ba785a2fa768a2c04d42b1b5fa46ff59f887363a1d8fdbaR1046), which is still needed.

My interpretation was to avoid a change in control-flow, so that all cases can be merged at the end. Or was something else the goal?

Reply:
I only suggested to follow the TruffleSqueak/Sulong loops as closely as possible. I do remember some peak performance issues in TruffleSqueak because it didn't do exactly what other Truffle languages were doing. Maybe that's part of the reason why your bytecode interpreters perform worse compared with their AST counterparts. But eventually, we also need to look into compilation traces and graphs.